File: MockLibAndroid/SmartMockLib/src/main/java/com/crescentflare/smartmock/utility/SmartMockPropertiesUtility.java
package com.crescentflare.smartmock.utility;

import android.content.Context;

import com.crescentflare.smartmock.model.SmartMockProperties;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.InputStream;

/**
 * Smart mock library utility: easily read and manage response properties
 */
public class SmartMockPropertiesUtility
{
    /**
     * Private constructor, only static methods allowed
     */

    private SmartMockPropertiesUtility()
    {
    }


    /**
     * Utility functions
     */

    public static SmartMockProperties readFile(Context context, String requestPath, String filePath)
    {
        SmartMockProperties properties = null;
        InputStream responseStream = SmartMockFileUtility.open(context, filePath + "/properties.json");
        if (responseStream != null)
        {
            String result = SmartMockFileUtility.readFromInputStream(responseStream);
            if (result != null)
            {
                try
                {
                    JSONObject propertiesJson = new JSONObject(result);
                    properties = new SmartMockProperties();
                    properties.parseJson(propertiesJson);
                    if (properties.getRedirect() != null)
                    {
                        SmartMockProperties redirectProperties = readFile(context, requestPath, filePath + "/" + properties.getRedirect());
                        if (redirectProperties != null)
                        {
                            redirectProperties.fallbackToProperties(properties);
                            properties = redirectProperties;
                        }
                    }
                }
                catch (JSONException ignored)
                {
                }
            }
        }
        if (properties == null)
        {
            properties = new SmartMockProperties();
        }
        properties.forceDefaults();
        return properties;
    }
}

File: MockLibIOS/SmartMockLib/Classes/utility/SmartMockPropertiesUtility.swift
//
//  SmartMockPropertiesUtility.swift
//  SmartMockServer Pod
//
//  Main library utility: easily read and manage response properties
//

class SmartMockPropertiesUtility {
    
    // --
    // MARK: Initialization
    // --
    
    private init() {
        // Private constructor, only static methods allowed
    }


    // --
    // MARK: Utility functions
    // --
    
    static func readFile(requestPath: String, filePath: String) -> SmartMockProperties {
        let properties = SmartMockProperties()
        if let responseStream = SmartMockFileUtility.open(path: filePath + "/properties.json") {
            let result = SmartMockFileUtility.readFromInputStream(responseStream)
            if let dictionary = SmartMockStringUtility.parseToDictionary(string: result) {
                properties.parseJsonDictionary(dictionary)
                if let redirect = properties.redirect {
                    let redirectProperties = readFile(requestPath: requestPath, filePath: filePath + "/" + redirect)
                    redirectProperties.fallbackTo(properties: properties)
                    redirectProperties.forceDefaults()
                    return redirectProperties
                }
            }
        }
        properties.forceDefaults()
        return properties
    }
    
}
